src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/ConcurrentQueueSemaphore.NetCoreApp.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    internal sealed partial class ConcurrentQueueSemaphore
+    {
+        public ValueTask WaitAsync(CancellationToken cancellationToken)
```

Comment:
I'm not sure what the point of returning ValueTask is - is this better than returning a Task? For the first path below, you can simply return Task.CompletedTask instead of `new ValueTask()`, and for the second one, just return `tcs.Task`...

Non-generic ValueTask is generally useful only in quite rare circumstances. It only really makes sense if it's going to be backed with IValueTaskSource (i.e. optimizing for asynchronous completion). It doesn't help for optimizing for synchronous completion, since you already have the option of returning Task.CompletedTask; that option doesn't exist in generic contexts, which is why `ValueTask<T>` is far more useful than non-generic `ValueTask`.